    ---
title: "GSEA"
output: html_notebook
---
https://github.com/saeyslab/muscatWrapper/blob/master/vignettes/basic_analysis_batchcor.md


# load

## packages
```{r}
setwd("/home/inf-38-2022/MScProject/Main_dir/")

library(Seurat)
library(SingleCellExperiment)
library(dplyr)
library(ggplot2)
library(muscatWrapper)

'%!in%' <- Negate(`%in%`)
set.seed(1234)
```


## data
```{r}
data <- readRDS("data/processed/DISCO_BM_v2_processed.rds")
```

Filtering
```{r}
simpl_ct_rm <- c("T/NK cells", "Stroma")
cycling_ct <- data@meta.data %>%
                filter(grepl("cycling", ct)) %>%
                distinct(ct) %>%
                pull(ct)
ct_rm <- c("Plasma cell", cycling_ct, "Erythrocyte", "Platelets", "Memory B cell", "Fetal HBG+ erythrocyte")
projects_keep <- c("GSE120221", "HCA_HematopoieticImmuneCellAtlas", "GSE185381")
data <- subset(data,
               subset = ct %!in% ct_rm & simplified_ct %!in% simpl_ct_rm & project_id %in% projects_keep)
```

```{r}
sce <- as.SingleCellExperiment(data)
```

# MUSCAT
```{r}
sample_id <- "sample"
group_id <- "Predicted_sex"
celltype_id <- "ct"
covariates <- "project_id"
```

## cell number
```{r}
# set minimum number of cells per sample and cell type to 10
min_cells <- 10

abundance_output <- muscatWrapper::get_abundance_info(sce, sample_id, group_id, celltype_id, min_cells, covariates = covariates)
head(abundance_output$abundance_data)

abundance_output$abund_plot_sample

```

## abundance between groups of interest
```{r}
abundance_output$abund_plot_group
abundance_output$abund_barplot
```

## statistical testing
```{r}
contrasts_oi <- c("'F-M','M-F'")
contrast_tbl <- tibble(contrast = c("F-M", "M-F"),
                        group = c("F", "M"))
```

```{r}
muscat_output <- muscatWrapper::muscat_analysis(
     sce = sce,
     celltype_id = celltype_id,
     sample_id = sample_id,
     group_id = group_id,
     covariates = covariates,
     contrasts_oi = contrasts_oi,
     contrast_tbl = contrast_tbl)
```

## check results
Save significant DEG
```{r}
muscat_output$celltype_de$celltype_de$de_output_tidy %>% filter(p_adj <= 0.05 & logFC >= 1) %>% arrange(p_adj) %>% filter(contrast == "M-F") -> DE_genes_table
write.table(DE_genes_table, "results/tables/DE_genes.csv", sep = ",", row.names = FALSE)
```


```{r}
muscat_output$celltype_info %>% lapply(head)
```
```{r}
muscat_output$celltype_de$celltype_de$de_output_tidy %>% arrange(p_adj) %>% head()
```
p-values
```{r}
muscat_output$celltype_de$hist_pvals
```


# Downstream
## exploring DE genes for a cell type
```{r}
celltype_oi <- "cDC1"
group_oi <- "M"

DE_genes <- muscat_output$celltype_de$celltype_de$de_output_tidy  %>% inner_join(contrast_tbl) %>% filter(group == group_oi) %>% filter(cluster_id == celltype_oi) %>% filter(p_adj <= 0.05 & logFC >= 1) %>% arrange(p_adj) %>% pull(gene) %>% unique()
DE_genes

```

```{r}
gene_oi <- DE_genes[1]

violin_plot <- muscatWrapper::make_DEgene_violin_plot(sce = sce, gene_oi = gene_oi, celltype_oi = celltype_oi, group_id = group_id, sample_id = sample_id, celltype_id = celltype_id, covariate_oi = covariates)
violin_plot
```

## GSEA

https://yulab-smu.top/biomedical-knowledge-mining-book/clusterprofiler-go.html
```{r}
library(clusterProfiler)
library(enrichplot)
library(ggplot2)
library(org.Hs.eg.db)
```

```{r}
# select genes from a cell type
d <- muscat_output$celltype_de$celltype_de$de_output_tidy %>%
                filter(cluster_id == "cDC1", contrast =="M-F" ) %>% # positive higher in males
                arrange(-logFC) %>%
                dplyr::select(gene, logFC)
geneList <- d$logFC
names(geneList) <- as.character(d$gene)
geneList <- sort(geneList, decreasing = TRUE)
```



```{r}
gse <- gseGO(geneList=geneList,
             ont ="ALL",
             keyType = "SYMBOL",
             minGSSize = 10,
             maxGSSize = 500,
             pvalueCutoff = 0.05,
             verbose = TRUE,
             OrgDb = org.Hs.eg.db,
             pAdjustMethod = "BH")
```

```{r}
dotplot(gse, color ="NES")
```


## compare cluster
make all the gene lists
```{r}

cell_types <- muscat_output$celltype_de$celltype_de$de_output_tidy$cluster_id %>% unique()
make_gene_list <- function (cell_type) {
    d <- muscat_output$celltype_de$celltype_de$de_output_tidy %>%
                    filter(cluster_id == !!cell_type , contrast =="M-F" ) %>% # positive higher in males
                    dplyr::select(gene, logFC)
    geneList <- d$logFC
    names(geneList) <- as.character(d$gene)
    geneList <- sort(geneList, decreasing = TRUE)
    return(geneList)
}


cluster_list <- lapply(cell_types, make_gene_list )
names(cluster_list) <- cell_types
#  save the list
date <- Sys.Date()
saveRDS(cluster_list, paste0("results/tables/cluster_list_", date, ".rds"))
```

```{r}
ck <- compareCluster(geneClusters = cluster_list,
                     fun = gseGO,
                     minGSSize = 10,
                     pvalueCutoff = 0.05,
                     pAdjustMethod = "BH",
                     OrgDb = org.Hs.eg.db,
                     ont = "ALL",
                     keyType = "SYMBOL")
date <- Sys.Date()

write.table(ck@compareClusterResult %>% filter(grouping!="Other"), paste0("results/tables/ck", date,".csv"), sep = ",", row.names = FALSE)

file.name <- paste0("04_GSEA/ck_", date, ".rds")
saveRDS(ck, file.name)
```





 * Mitochondrial
    * Mitochondrial protein-containing complex
    * Mitochondrial translation
    * Mitochondrial inner membrane
    * Mitochondrial matrix
    * Mitochondrial gene expression
    * Mitochondrial envelope
* Cellular machinery
    * Organelle inner membrane
    * Mitochondrial membrane
    * Electron transport chain
    * RNA polymerase II cis-regulatory region sequence-specific DNA binding
    * Cis-regulatory region sequence-specific DNA binding
    * Transcription cis-regulatory region binding
    * Sequence-specific double-stranded DNA binding
* Cell processes
    * Chromosome organization
    * Cell cycle process
    * DNA replication
    * DNA metabolic process
    * Mitochondrion
    * Translation
    * Cellular macromolecule biosynthetic process
    * Amide biosynthetic process
    * Peptide metabolic process
    * Ribonucleoprotein complex
```{r}
# x_order <- c( "Myelocyte", "S100A+ preNeutrophil", "CD14+MHCIIhigh monocyte", "CD14+MHCIIlow monocyte", "CD16+ monocyte", "cDC1", "cDC2",  "pDC",  "proB cell","preB cell","Naive B cell")

x_order <- c( "S100A+ preNeutrophil","CD14+MHCIIlow monocyte",  "CD14+MHCIIhigh monocyte",  "CD16+ monocyte", "cDC1", "cDC2", "proB cell","preB cell","Naive B cell")


Mitochondrial <- c("mitochondrial protein-containing complex",
                   "mitochondrial ribosome",
                   "mitochondrial translation",
                   "mitochondrial inner membrane",
                   "organelle inner membrane",
                   "mitochondrial matrix",
                   "mitochondrial envelope",
                   "mitochondrion",
                   "electron transport chain",
                   "respiratory electron transport chain",
                   "organellar ribosome")

GeneticRegulation <- c("cis-regulatory region sequence-specific DNA binding",
                       "RNA polymerase II transcription regulatory region sequence-specific DNA binding",
                       "sequence-specific double-stranded DNA binding",
                       "transcription regulatory region nucleic acid binding",
                       "sequence-specific DNA binding")

CellCycle <- c("cell cycle process",
                       "chromosome organization",
                       "DNA replication",
                       "DNA metabolic process",
                       "DNA-templated DNA replication")

RibosomalFunctions <- c(
                        "ribosome",
                        "ribosomal subunit",
                        "cytosolic ribosome",
                        "translation",
                        "structural constituent of ribosome")

ck@compareClusterResult$grouping <- ifelse(ck@compareClusterResult$Description %in% Mitochondrial, "Mitochondrial Function",
                                           ifelse(ck@compareClusterResult$Description %in% GeneticRegulation, "Gene Regulation",
                                                  ifelse(ck@compareClusterResult$Description %in% CellCycle, "Cell Cycle",
                                                         ifelse(ck@compareClusterResult$Description %in% RibosomalFunctions, "Ribosomal Function", "Other"))))




ck %>%
  filter(grouping != "Other") %>%
  dotplot(,showCategory=5, font.size = 15, label_format = 55, color = "NES", size = "setSize")  -> dotplot_clusters
dotplot_clusters <- dotplot_clusters + theme(axis.text.x = element_text(angle = 20, hjust=1))
# change the color scale to nes_cat
dotplot_clusters <- dotplot_clusters + scale_color_gradient2(low = "#8A69D4", high = "#FFD635" , midpoint = 0, mid = "gray") +
  facet_grid( factor(grouping, levels = c("Mitochondrial Function", "Cell Cycle", "Gene Regulation", "Ribosomal Function"))~.,
              scales = "free_y", switch = "y", space = "free",
  labeller = label_wrap_gen(width = 10)) +
  labs(y=NULL, x ="Cell type", size = "Set size") +
    theme(strip.background = element_blank(), strip.placement = "outside",
          strip.text.y = ggtext::element_textbox_simple(size =15,  face = "bold"),
          axis.text.y = element_text(size = 10)) +

  scale_x_discrete(limits = x_order)
  #   scale_y_discrete(limits = rev(y_order))
dotplot_clusters + patchwork::plot_annotation(caption = "Figure 6. Females overexpress mitochondrial-related genes, whereas males show enrichment in gene regulation pathways. GSEA for GO categories. Only cell types with at least one signficant category (p.adj > 0.5) are shown. Size represents the number of genes in the category and color represents the NES (Overexpressed in males (gold) or females (purple)). GO categories have been grouped into general terms.") &
 theme(plot.caption = ggtext::element_textbox_simple(size=11))-> dotplot_clusters
ggsave("results/plots/fig6_dotplot_clusters_gsea.png", dotplot_clusters, width = 14, height = 9, units = "in")
# with date
ggsave(paste0("results/plots/fig6_dotplot_clusters_gsea_", date, ".png"), dotplot_clusters, width = 14, height = 9, units = "in")
saveRDS(dotplot_clusters, "results/plots/fig6_dotplot_clusters_gsea.rds")
```



## KEGG
Convert SYMBOL to ENTREZID. SOME fail. Around 5% of genes
```{r}

cluster_list_kegg <- lapply(cluster_list, function(x) {

  name_table <- bitr(names(x), fromType = "SYMBOL", toType = "UNIPROT",
            OrgDb = org.Hs.eg.db) %>% distinct(SYMBOL, .keep_all = TRUE)

  names(x) <- name_table$UNIPROT
  # remove NA
  x <- x[!is.na(names(x))]
  return(x)
})
Sys.Date() -> date
saveRDS(cluster_list_kegg, paste0("results/tables/cluster_list_kegg_", date, ".rds"))
```




```{r}
ck_kegg <- compareCluster(geneClusters = cluster_list_kegg,
                     fun = gseKEGG,
                     minGSSize = 10,
                     pvalueCutoff = 0.05,
                     pAdjustMethod = "BH",
                     organism ="hsa",
                     keyType = "uniprot"
    )
# save
date <- Sys.Date()
file.name <- paste0("results/tables/ck_kegg_", date, ".rds")
saveRDS(ck_kegg, file.name)
```

dotplot
```{r}
x_order <- c("Promyelocyte", "Myelocyte", "S100A+ preNeutrophil","CD14+MHCIIlow monocyte",  "CD14+MHCIIhigh monocyte",  "CD16+ monocyte", "cDC1", "pDC","preB cell","Naive B cell")


# Disease category
Disease <- c( "Coronavirus disease - COVID-19", "Parkinson disease", "Prion disease", "Alzheimer disease", "Huntington disease", "Pathways of neurodegeneration - multiple diseases", "Non-alcoholic fatty liver disease", "Amyotrophic lateral sclerosis", "Retrograde endocannabinoid signaling", "Chemical carcinogenesis - reactive oxygen species", "Diabetic cardiomyopathy")

# Metabolic category
Metabolic <- c("Oxidative phosphorylation","Carbon metabolism",  "Metabolic pathways", "Thermogenesis")

DNA <- c( "DNA replication", "Mismatch repair")

ck_kegg@compareClusterResult$grouping <- ifelse(ck_kegg@compareClusterResult$Description %in% Disease, "Disease",
                                           ifelse(ck_kegg@compareClusterResult$Description %in% Metabolic, "Metabolic",
                                                  ifelse(ck_kegg@compareClusterResult$Description %in% DNA, "DNA", "Other")))



ck_kegg %>%
  filter(grouping != "Other") %>%
  dotplot(,showCategory=5, font.size = 15, label_format = 55, color = "NES", size = "setSize")  + theme(axis.text.x = element_text(angle = 20, hjust=1)) -> dotplot_clusters_kegg
# change the color scale to nes_cat
dotplot_clusters_kegg <- dotplot_clusters_kegg + scale_color_gradient2(low = "#8A69D4", high = "#FFD635" , midpoint = 0, mid = "gray") +
  facet_grid( factor(grouping, levels = c("Metabolic", "Disease", "DNA"))~.,
              scales = "free_y", switch = "y", space = "free",
  labeller = label_wrap_gen(width = 10)) +
  labs(y=NULL, x ="Cell type", size = "Set size") +
    theme(strip.background = element_blank(), strip.placement = "outside",
          strip.text.y = ggtext::element_textbox_simple(size =15,  face = "bold"),
          axis.text.y = element_text(size = 10)) +

  scale_x_discrete(limits = x_order)


# dotplot_clusters + patchwork::plot_annotation(caption = "Figure 6. Females overexpress mitochondrial-related genes, whereas males show enrichment in gene regulation pathways. GSEA for GO categories. Only cell types with at least one signficant category (p.adj > 0.5) are shown. Size represents the number of genes in the category and color represents the NES (Overexpressed in males (gold) or females (purple)). GO categories have been grouped into general terms.") &
#  theme(plot.caption = ggtext::element_textbox_simple(size=11))-> dotplot_clusters
# ggsave("results/plots/fig6_dotplot_clusters_gsea.png", dotplot_clusters, width = 14, height = 9, units = "in")
# # with date
date <- Sys.Date()
ggsave(paste0("results/plots/fig7_dotplot_kegg_clusters_gsea_", date, ".png"), dotplot_clusters_kegg, width = 14, height = 9, units = "in")
saveRDS(dotplot_clusters_kegg, "results/plots/fig7_dotplot_kegg_clusters_gsea.rds")
```

## heatmap

```{r}
library(clusterProfiler)
library(dplyr)
library(ggplot2)
library(tidyr)
library(org.Hs.eg.db)
library(Seurat)
x_y_heatmap <- function (ck, category , cell_type, seurat_obj = data, sample_col = "sample", sex_col = "Predicted_sex" ,assay="RNA", project_col ="project_id") {
  # get genes to plot
  ck@compareClusterResult %>%
    filter(Description == category, Cluster ==cell_type) %>%
    pull(core_enrichment) %>%
    strsplit(split = "/")  %>%
    unlist()-> genes_to_plot
  # get gene names in symbol
    bitr(genes_to_plot, fromType = "UNIPROT", toType="SYMBOL", OrgDb= org.Hs.eg.db) -> genes_to_plot
  # get average expression per sample
  sample.averages <- AverageExpression(seurat_obj, features = genes_to_plot$SYMBOL,
                                       return.seurat = TRUE, group.by = sample_col)
  Idents(sample.averages) <- names(Idents(sample.averages)) # idk why i do this
  mean_gene_expr <- rowSums(sample.averages[[assay]]@scale.data)
  var_genes <- names(mean_gene_expr[mean_gene_expr != 0])

  # get M - F sample order for heatmap
    seurat_obj@meta.data %>%
      # arrange(!!sym(sex_col)) %>%
      distinct(!!sym(sample_col), !!sym(sex_col), !!sym(project_col), age) -> sample_order
  # Idents(sample.averages) <- factor(sample.averages@active.ident, levels = sample_order$sample_col)
  # Add idents as metadata
    sample.averages <- AddMetaData(sample.averages, metadata = Idents(sample.averages),
                                     col.name = sample_col)
  sample_order %>%
    right_join(sample.averages@meta.data, by = c("sample" = "sample")) -> sample_order
    rownames(sample_order) <- sample_order$sample
   # add Predicted_sex as metadata
  sample.averages <- AddMetaData(sample.averages, metadata = sample_order[,2],
                                 col.name = "sex")
  sample.averages <- AddMetaData(sample.averages, metadata = sample_order[,3],
                                 col.name = "project_id")
    sample.averages <- AddMetaData(sample.averages, metadata = sample_order[,4],
                                 col.name = "age")
  # both sex-project_id
  sample.averages <- AddMetaData(sample.averages, metadata = paste(sample_order[,3], sample_order[,2], sep="_"), col.name = "sex_proj")
  levels <- sample.averages@meta.data %>% arrange(sex_proj) %>% distinct(sex_proj) %>% pull(sex_proj)
  Idents(sample.averages) <- factor(sample.averages@meta.data$sex_proj, levels = levels )
  # plot heatmap
    DoHeatmap(sample.averages, features = var_genes, size = 2, draw.lines = FALSE, group.by = "sex_proj",
              assay = assay, disp.max = 2, disp.min = -0.8)  + NoLegend()-> x_y_heatmap_plot
  # make boxplot
  sample.averages[[assay]]@scale.data %>%
    as.data.frame() %>%
    mutate(gene =rownames(.)) %>%
    filter(gene %in% var_genes) %>%
    pivot_longer(-gene, names_to = "sample", values_to = "expr") %>%
    left_join(sample_order, by = "sample") %>%
    mutate(sample = factor(sample, levels = sample_order$sample_col)) %>%
    ggplot(aes(x = gene, y = expr, fill = !!sym(sex_col))) +
    facet_wrap(vars(project_id)) +
    geom_boxplot() -> boxplot
    return(list(x_y_heatmap_plot, boxplot, levels, sample_order, sample.averages@meta.data))
}
plot <- x_y_heatmap(ck_kegg, "Oxidative phosphorylation", "S100A+ preNeutrophil")
```

logFC dotplot
```{r}
category <- "Oxidative phosphorylation"
cell_type <- "S100A+ preNeutrophil"
ck_kegg@compareClusterResult %>%
    filter(Description == category, Cluster ==cell_type) %>%
    pull(core_enrichment) %>%
    strsplit(split = "/")  %>%
    unlist()-> genes_to_plot
  # get gene names in symbol
    bitr(genes_to_plot, fromType = "UNIPROT", toType="SYMBOL", OrgDb= org.Hs.eg.db) -> genes_to_plot

muscat_output$celltype_de$celltype_de$de_output_tidy %>%
  filter(contrast == "M-F") %>%
  filter(gene %in% genes_to_plot$SYMBOL) %>%
  ggplot(aes(x = logFC, y = gene, color = cluster_id)) +
    geom_point()
```



## pathview
```{r}
library(pathview)
library(clusterProfiler)
date <- Sys.Date()
#create dir for pathways
dir.create("results/pathways", showWarnings = FALSE)
dir.create("results/pathways/kegg", showWarnings = FALSE)
```

```{r}
cell_type <- "S100A+ preNeutrophil"
# get the gene list
gene_list <- cluster_list_kegg[[cell_type]]
a <- bitr_kegg(names(gene_list), fromType = "uniprot", toType="kegg", organism="hsa")
names(gene_list) <- a$kegg # ther will be NAs
```

Create the pathway
```{r}
pathway <- "hsa01100"
setwd("results/pathways/kegg")
pathway_obj <- pathview(gene.data  = gene_list,
                     pathway.id = pathway,
                     species    = "hsa",
                     limit      = list(gene=1, cpd=1),
                        out.suffix = paste0(cell_type, "_", pathway, date)
)
setwd("/home/inf-38-2022/MScProject/Main_dir/")
```

```{r}
```

```{r}